src/orange/orangeinp/ConvexRegion.cc

```diff
@@ -368,6 +368,31 @@ GenTrap::GenTrap(real_type halfz, VecReal2 const& lo, VecReal2 const& hi)
     }
 }
 
+/*!
+ * Construct a TRD from 5 half-lengths: one half-height, {hx1,hy1} for a in -z,
+ * and {hx2,hy2} in +z
+ */
+GenTrap::GenTrap(real_type halfz, Real2 const& hxy1, Real2 const& hxy2)
+    : hz_{halfz}, lo_{4}, hi_{4}
+{
+    CELER_VALIDATE(hxy1[0] > 0 && hxy2[0] > 0,
+                   << "nonpositive x half-length: " << hxy1[0] << " "
+                   << hxy2[0]);
+    CELER_VALIDATE(hxy1[1] > 0 && hxy2[1] > 0,
+                   << "nonpositive y half-length: " << hxy1[1] << " "
+                   << hxy2[1]);
+    CELER_VALIDATE(hz_ > 0, << "nonpositive halfheight: " << hz_);
+
+    lo_[0] = GenTrap::Real2{-hxy1[0], -hxy1[1]};
+    lo_[1] = GenTrap::Real2{+hxy1[0], -hxy1[1]};
+    lo_[2] = GenTrap::Real2{+hxy1[0], +hxy1[1]};
+    lo_[3] = GenTrap::Real2{-hxy1[0], +hxy1[1]};
+    hi_[0] = GenTrap::Real2{-hxy2[0], -hxy2[1]};
+    hi_[1] = GenTrap::Real2{+hxy2[0], -hxy2[1]};
+    hi_[2] = GenTrap::Real2{+hxy2[0], +hxy2[1]};
+    hi_[3] = GenTrap::Real2{-hxy2[0], +hxy2[1]};
```

Comment:
Since this code is inside the class now, you don't have to do `GenTrap::`, and in fact you don't even need the `::Real2` since it's explicit.

I think the code would be a bit safer if instead of sizing it on 376 and writing directly to the elements, you leave off the `lo_` and `hi_` and write this as an initializer list:
```suggestion
    lo_ = {{-hxy1[0], -hxy1[1]},
           {+hxy1[0], -hxy1[1]},
           {+hxy1[0], +hxy1[1]},
           {-hxy1[0], +hxy1[1]}};
    hi_ = {{-hxy2[0], -hxy2[1]},
           {+hxy2[0], -hxy2[1]},
           {+hxy2[0], +hxy2[1]},
           {-hxy2[0], +hxy2[1]}};
```

I'm actually not even sure if `, lo_{4}, hi_{4}` in your constructor is *resizing* it to 4 or *initializing* it with an initializer list... the fact that `vector` can be constructed with `std::initializer_list` changes the semantics in weird ways...